line_follower.py

```diff
@@ -0,0 +1,218 @@
+                         next_direction = Direction.down
+                     else:
+                         if self.direction == Direction.up:
+                            next_direction = Direction.left
+                 
+                 if avgBigX > avgSmallX and avgBigY > avgSmallY:
+                     if self.direction == Direction.right:
+                         next_direction = Direction.up
+                     else:
+                         if self.direction == Direction.down:
+                            next_direction = Direction.left
+                 
+                 if avgBigX < avgSmallX and avgBigY > avgSmallY:
+                     if self.direction == Direction.left:
+                         next_direction = Direction.up
+                     else:
+                         if self.direction == Direction.down:
+                            next_direction = Direction.right
+
+                 # change the count based on found direction and change direction if count equals zero
+                 if self.immediate_dir == next_direction and next_direction != Direction.neutral:
+                     self.change_dir_count -= 1
+                 else:
+                    if self.change_dir_count < self.turns_needed:
+                        self.change_dir_count += 1
+                 if self.change_dir_count == self.turns_needed:
+                    self.immediate_dir = next_direction
+                 if self.change_dir_count == 0:
+                     if self.immediate_dir != Direction.neutral:
+                         self.change_dir_count = self.turns_needed 
+                         self.next_dir = self.immediate_dir
+                         self.find_turn = False
+                         self.turn = current_turn
+            else:
+                
+                # determine whether there is a turn based on hull/bounding rectabgle of past few images
+                self.dir_change.append(current_turn)
+                if len(self.dir_change) > self.dir_change_queue_length:
+                    self.dir_change.pop(0)
+                turn_avg = sum(self.dir_change)/len(self.dir_change)
+                if turn_avg > self.change_turn_multiplier * self.turn:
+                    self.find_turn = True
+                else:
+                    if turn_avg < self.adopt_new_turn_multiplier * self.turn:
+                        self.direction = self.next_dir
+                    if turn_avg < self.decrease_turn_multiplier * self.turn:
+                        self.turn = turn_avg
+                        
+                #draw hull, contour, and bouding rectangle for testing purposes
+                cv2.drawContours(img, [hull], 0, (0, 255, 0), 2)
+                cv2.drawContours(img, [cnt], 0, (0, 255, 0), 2)
+                cv2.drawContours(img, [boundingRect], 0, (255, 0, 0), 2)
+        else:
+            print('No red line found')
+
+        cv2.imshow('img', img)
+        cv2.waitKey(1)
+
+
```

Comment:
wrap this in
```python
if __name__ == "__main__"
```

if someone imports this file, then it will run these lines of code.  the if block prevents that and only runs them if you actually run this file